nebula_hw_interfaces/include/nebula_hw_interfaces/nebula_hw_interfaces_common/connections/udp.hpp

```diff
@@ -0,0 +1,383 @@
+// Copyright 2024 TIER IV, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <nebula_common/util/expected.hpp>
+
+#include <arpa/inet.h>
+#include <netinet/in.h>
+#include <sys/poll.h>
+#include <sys/socket.h>
+#include <unistd.h>
+
+#include <algorithm>
+#include <atomic>
+#include <cassert>
+#include <cerrno>
+#include <cstddef>
+#include <cstdint>
+#include <cstring>
+#include <exception>
+#include <functional>
+#include <optional>
+#include <stdexcept>
+#include <string>
+#include <thread>
+#include <utility>
+#include <vector>
+
+namespace nebula::drivers::connections
+{
+
+class SocketError : public std::exception
+{
+public:
+  explicit SocketError(int err_no) : what_(strerror(err_no)) {}
+
+  explicit SocketError(const char * msg) : what_(msg) {}
+
+  const char * what() const noexcept override { return what_.c_str(); }
+
+private:
+  std::string what_;
+};
+
+class UsageError : public std::runtime_error
+{
+public:
+  explicit UsageError(const std::string & msg) : std::runtime_error(msg) {}
+};
+
+class UdpSocket
+{
+private:
+  struct Endpoint
+  {
+    std::string ip;
+    uint16_t port;
+  };
+
+  struct MsgBuffers
+  {
+    msghdr msg{};
+    iovec iov{};
+    std::array<std::byte, 1024> control;
+    sockaddr_in sender_addr;
+  };
+
+  class DropMonitor
+  {
+    uint32_t last_drop_counter_{0};
+
+  public:
+    uint32_t get_drops_since_last_receive(uint32_t current_drop_counter)
+    {
+      uint32_t last = last_drop_counter_;
+      last_drop_counter_ = current_drop_counter;
+
+      bool counter_did_wrap = current_drop_counter < last;
+      if (counter_did_wrap) {
+        return (UINT32_MAX - last) + current_drop_counter;
+      }
+
+      return current_drop_counter - last;
+    }
+  };
+
+  enum class State { UNINITIALIZED, INITIALIZED, BOUND, ACTIVE };
+
+  static const int g_poll_timeout_ms = 10;
```

Comment:
This is a tricky one: the timeout should be long enough to definitely receive at least one packet during normal operation, and short enough that when Nebula is shutdown (e.g. via `CTRL+C`), the user does not have to wait for a noticeable amount of time.
While I agree that this is a bit of an arbitrary value, having it as a parameter will lead people to abuse it (e.g. set it to 10s and then complain that Nebula takes forever to exit).
We could have it as a constructor argument, and have Nebula set that to a constant defined in `nebula_common` with proper documentation.

Reply:
I don't mean a parameter that is exposed to say, ROS users - more a parameter in the UDP functionality. So for example, someone integrating a particularly obnoxious sensor could adjust it if required.
That said, if it is only really useful in tuning for network issues, it's a bit harder to see a use case for it without opening access to the user (rather than programmer) side.

Reply:
Addressed in 11ed243.